File: ReOrderLogFile.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 14 16:02:34 2019

@author: tanvirkaur
"""
#Leetcode accceptance = yes
# Time complexity = O(n) where n is the total length of the logs

class Solution(object):
    def reorderLogFiles(self, logs):
        """
        :type logs: List[str]
        :rtype: List[str]
        """
        result = []
        digit1 = []
        letter = []
        for log in logs:
            word = log.split()
            if word[1].isdigit():
                digit1.append(word)
            else:
                letter.append(word)
        #letter.sort(cmp = lambda x,y: x[1]-y[1] if x[1]!= y[1]! else x[0]-y[0])
        #return " ".join(letter) + digit
        return [" ".join(i) for i in sorted(letter, key = lambda x: x[1:]+[x[0]]) + digit1]

File: ReorderLogFile.java
class ReorderLogFile {
  /**
   * Time Complexity: O(n m logn )
   * 
   * Space Complexity: O(mlogn)
   *   m -> average word length in the strings
   * 
   * Were you able to solve this on leetcode? Yes.
   */
  public String[] reorderLogFiles(String[] logs) {
    Arrays.sort(logs, (str1, str2) -> {
        
        String[] strArr1 = str1.split(" ", 2);
        String[] strArr2 = str2.split(" ", 2);
        
        boolean str1IsDigit = Character.isDigit(strArr1[1].charAt(0));
        boolean str2IsDigit = Character.isDigit(strArr2[1].charAt(0));
        
        if(!str1IsDigit && !str2IsDigit) {
            // both are Strings
            
            if(strArr1[1].equals(strArr2[1])) {
                return strArr1[0].compareTo(strArr2[0]);
            }
            
            return strArr1[1].compareTo(strArr2[1]);
            
        } else if(!str1IsDigit && str2IsDigit) {
            return -1;
        } else if(str1IsDigit && !str2IsDigit) {
            return 1;
        } else {
            return 0;
        }

    });
    
    return logs;
}
}